sklearn/metrics/_classification.py

```diff
@@ -2442,9 +2454,16 @@ def balanced_accuracy_score(y_true, y_pred, *, sample_weight=None, adjusted=Fals
     C = confusion_matrix(y_true, y_pred, sample_weight=sample_weight)
     with np.errstate(divide="ignore", invalid="ignore"):
         per_class = np.diag(C) / C.sum(axis=1)
-    if np.any(np.isnan(per_class)):
-        warnings.warn("y_pred contains classes not in y_true")
-        per_class = per_class[~np.isnan(per_class)]
+    if np.any(np.isnan(per_class)) or np.any(C.sum(axis=1) == 0):
+        if zero_division == "warn":
+            warnings.warn(
+                "y_pred contains classes not in y_true or "
+                "some classes have no true samples."
+            )
```

Comment:
Since we compute the diagonal of C dividive by the sum of the row of C, the only exception is a "0 / 0" leading to a `nan`. In this case, it means that we don't have any `y_true` value for the associated class but `np.unique` provide such class so it means that it was present in `y_pred`.

So the warning was correct and I think that your changes state twice the same fact.


```suggestion
    if np.any(np.isnan(per_class)):
        if zero_division == "warn":
            warnings.warn("y_pred contains classes not in y_true")
```